### ISSUE TITLE
Secure Memory Erasing for SPIKE Nexus and SPIKE Keeper

### ISSUE THREAD

**Issue by v0lkan** [spiffe/spike#68]
Both SPIKE Nexus and SPIKE keeper (albeit temporarily) store root keys and Shamir Shards that they don't directly need.

Although these are deleted when no longer needed, it is up to Go garbage collector "when" to delete them.

Instead, we need to zero out unused data before gc.

According to NIST SP 800-88 Rev. 1 (Guidelines for Media Sanitization), a single overwrite pass with zeros is sufficient for modern storage devices. And since our data is in memory, one pass should be enough too. Multiple pass overwrites (like Gutmann's 35-pass method) are no longer considered necessary (due to high storage densities in modern devices).

The key is ensuring the overwrite actually occurs and isn't optimized away by the compiler, which the previous code handles correctly.

Here is one way to do it:

```
func Clear[T any](s *T) {
    p := unsafe.Pointer(s)
    size := unsafe.Sizeof(*s)
    b := (*[1 << 30]byte)(p)[:size:size]
    runtime.KeepAlive(s) // make sure the data is actually wiped before gc has time to interfere.
    for i := range b {
        b[i] = 0
    }
}

// Usage
type SensitiveData struct {
    Key    [32]byte
    Token  string
    UserId int64
}

func main() {
    data := &SensitiveData{...}
    defer Clear(data)
    // Use data
}
```

**Comment by kfox1111** [comment]
It would be interesting to try and chat with some of the folks under the cncf that are involved with confidential computing... Maybe there is an existing, nice interface we could reuse that keeps the keys in protected memory, and/or has nice wiping features.

**Comment by v0lkan** [comment]
> It would be interesting to try and chat with some of the folks under the cncf

That's a good idea indeed; I'm noting it down.

**Comment by dsykes16** [comment]
Does Go provide any assurances that the value pointed to is the _only_ copy it ever put on the heap?

https://www.gnu.org/software/libc/manual/html_node/Memory-Protection.html
https://pkg.go.dev/golang.org/x/sys/unix?utm_source=godoc#Mprotect

It looks like the required syscalls are available in Go.

**Comment by gurkanguray** [comment]
I've implemented the secure memory erasing solution in PR #107 . Following @v0lkan 's recommendation on the issue, I opted for the memory zeroing approach rather than the mprotect-based solution. Mprotect is UNIX-specific and would require different implementations for Windows and other platforms. Since the implemented approach already aligns with NIST guidelines for clearing sensitive data, I think it's good to go with that. However, if we need memory protection with mprotect in the future, then I would be happy to work on this part

**Comment by v0lkan** [comment]
The implementation in PR #107 is an improvement over what we have ( i.e. no secure memory erasure whatsoever :) ).

But I do agree with @kfox1111 that it is a good idea to pick the brains of other subject-matter experts too.

That may answer question about mrpotect along with other question like:
* Is zeroing out once good enough, or shall we do it multiple times?
* would writing random garbage several times better than simply zeroing out?

--

As an aside, [SPIKE's security model](https://spike.ist/architecture/security-model/) does not guarantee any protection against memory-based attacks. That being said, security is a layered cheese, and if we can provide protection against certain attack vectors, we should.


--

@kfox1111 do you know anyone that you can introduce in CNCF? 

--


I'll check out the PR too, to see if there's anything that might be missing.

**Comment by v0lkan** [comment]
@dsykes16 @sahinakyol 

> Does Go provide any assurances that the value pointed to is the only copy it ever put on the heap?

To my knowledge,Go's memory model does not provide a strict guarantee that an allocated value on the heap will be the only instance of that value in memory. 

There are multiple scenarios where the Go compiler or runtime may introduce additional copies:

* Escape Analysis and Stack to Heap Promotion: If a variable escapes the function scope, Go’s compiler moves it from stack to heap. The original stack-allocated copy may be discarded, but a duplicate might temporarily exist.
* Garbage Collector and Compaction: Go’s garbage collector may move objects in memory under certain conditions. However, further research persuaded me that since Go does not use a compacting garbage collector, this is rare (again, better to consult a SME on that).
* Pass-by-Value Semantics: Go passes function arguments by value (except slices, maps, and pointers, which are references). If you assign a struct directly, a copy is made.
* Slice Reslicing and Growth: While slices are references to an underlying array, modifying a slice (such as appending elements) can trigger a reallocation, resulting in copies of the original data.
* Compiler Optimizations: The Go compiler may introduce copies during register allocation, inlining, or other optimizations that the developer cannot control.

^ My best bet is; we can zero out what we can and assume that the compiler is not too crazy to do too unexpected things.

The linked PR

* erases the memory of a slice using unsafe Go code; and since the slice is practically a pointer type, I'd assume it will zero out everything it points to, For the slice, while the slice "header" is passed by value, the slice itself (i.e. the C…ish "array", in a sense) holds the slice’s elements remains in the same memory location as long as the slice is not resized (i.e., as long as append does not trigger a reallocation).
* eraases the contents of a global byte slice (rootKey) the same way.

In both cases, I believe it will do the right thing.

And in both cases, I believe we can (mostly) safely assume that there is a single copy.

Since
* in the `rootKey` case, it's a shared global variable.
* an in the "shard erasure" case, it's a local slice will be created on the stack and will remain on the stack as it never escapes the function.

I'll need to look at the code better. 

But
1. it's still an improvement.
2. I'd love to have other paranoid eyes look at it as well :) .

**Comment by v0lkan** [comment]
Assigned to @gurkanguray since he owns the related PR too.

**Comment by v0lkan** [comment]
TL;DR: version of the above.

* We wipe `rootKey`. -- `rootKey` is a heap-allocated slice of bytes. Unless the compiler is utterly mad, there will be a single copy of it on the heap at all times.
* We wipe decoded shards (which are byte slices too). These shards are local variables in a function body. They are (my understanding) stack-allocated. They don't escape the function so, again, unless the compiler is utterly mad, there is a single copy of those, and the "secure erase" (TM) will erase those too.

^ So that's a good enough method for the two particular variables, and for our particular purpose (I think)

---

Is it the "best" way to do this? Is there a library that we can call and not have to do this heap/stack mental excercise? can entropy be reversed?

^ those are the questions that we should discuss with CNCF Confidential Computing experts.

**Comment by v0lkan** [comment]
To make this even more deterministic, we may initially set root key to be 32 bytes (of zeroes) instead of letting it dynamically grow. Maybe that can be a separate, follow-up issue (as it will require us to refactor a bunch of nil-checks against the root key too)

**Comment by v0lkan** [comment]
More on memory optimization (this is ChatGPT talking so keep your salt-shakers with you, but it made sense to me)

> The Go compiler performs several optimizations, but none of these will duplicate your data unexpectedly; in fact, they usually aim to eliminate unnecessary work.
>
> For instance, escape analysis is itself an optimization to avoid heap allocations when not needed, and it can even avoid copy operations by keeping data on the stack. 
>
> The compiler also has optimizations in its SSA phase that can eliminate redundant assignments or copy operations. For example, when you return a large struct from a function, the compiler can optimize away intermediate copies by constructing the return value directly in the caller’s stack frame (this is a form of return value optimization in Go’s calling convention). 
>
> These optimizations are all under the hood but preserve the language’s semantics – meaning the behavior you see in Go code corresponds to what you’d expect from the language spec, just done efficiently.

**Comment by v0lkan** [comment]
Another thing that can be done (for extra security) would be to use a fixed width (32byte) array for root key, instead of a slice (`var rootKey [32]byte`)

* As a global variable, this will be stored in the .data segment (if initialized with a value).
* It will no move around dynamically like heap-allocated memory.
* This will also avoid unintended copies that could occur with slices.

One obvious disadvantage would be that the root key will be fixed-length.
Which is okay for now.

I'll create a separate issue about this.

**Comment by v0lkan** [comment]
Closing this as we've merged the related PR.

However, I've taken a note to follow up this with CNCF folks too.

**Comment by gurkanguray** [comment]
Thank you for the comments 🙌🏻

**Comment by v0lkan** [comment]
additional comments.

wrt `mlock`; 
* it requires root privileges or `CAP_IPC_LOCK` in Linux.
* there is no windows equivalent.

however without `mlock` if swap is on, zeroing out memory won't be good enough as we won't have any control ove whatever is written on swap.

with **swap off** my research so far suggests that the lined PR will save the day.

--

@kfox1111 has suggested to add "turning swap off" to the [production guide](https://spike.ist/operations/production/) as a recommendation. 

Turning swap off is recommended for bare-metal k8s control plane and worker nodes anyway; so chances are, it will be turned off if a modern containerized (doker, k8s, etc) environment.

If turning swap is not an option, encrypting swap could be another option too.


---

Also we can either "try and shrug" using `mlock` as in

```go
	err := unix.Mlock(data)
	if err != nil {
		// do nothing
	}

	// Zero out memory
	for i := range data {
		data[i] = 0
	}
```

Or we can put the mlock option behind a feature flag.

If the user has granted `CAP_IPC_LOCK` to SPIKE Nexus, then they can also set something like `SPIKE_NEXUS_USE_MLOCK="true"` as an env var, and try locking memory if the variable is set (will be "false" by default).

I'll create seperate issues for these.
